File: lintcode/remove-element.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/remove-element/

class Solution:
    """
    @param A: A list of integers
    @param elem: An integer
    @return: The new length after remove
    """
    def removeElement(self, A, elem):
        # write your code here
        cursor = 0
        reader = 0
        while reader < len(A):
            while reader < len(A) and A[reader] == elem:
                reader += 1
            if reader < len(A):
                A[cursor] = A[reader]
                cursor += 1
                reader += 1

        return cursor

if __name__ == '__main__':
    s = Solution()
    array = [0,4,4,0,0,2,4,4]
    new_len = s.removeElement(array, 4)
    print new_len, array

File: leetcode/remove-element.cc
// Question: https://leetcode.com/problems/remove-element/

#include "headers.h"

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int i = 0;
        int j = 0;
        while (j < nums.size()) {
            if (nums[j] != val) {
                swap(nums[i], nums[j]);
                ++i;
            }
            ++j;
        }
        return i;
    }

private:
    void swap(int& x, int& y) {
        int t = x;
        x = y;
        y = t;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> nums{1,2,1,3,1,4,1,5};
    cout << s.removeElement(nums, 1) << endl;
    return 0;
}
